pageserver/src/layered_repository/layer_map.rs

```diff
@@ -294,8 +297,13 @@ impl SegEntry {
         false
     }
 
-    pub fn insert_open(&mut self, layer: Arc<InMemoryLayer>) {
-        assert!(self.open.is_none());
+    // Set new open layer for a SegEntry.
+    // It's ok to rewrite previous open layer,
+    // but only if it is not writeable anymore.
+    pub fn update_open(&mut self, layer: Arc<InMemoryLayer>) {
+        if let Some(prev_open) = &self.open {
+            assert!(!prev_open.is_writeable());
+        }
         self.open = Some(layer);
```

Comment:
> Allowing replacing the open layer seems a bit sloppy to me. Previously, it was clear that each segment would have only one open layer at a time, now the rule is more complicated.

Do you see any issues with correctness here?
We should definitely use better terminology here because now it is quite confusing.
I think if we rename `open_layers` to `inmemory_layers`, things will be more clear - multiple in-memory layers can exist, but only one of them is open (aka writeable).
But I'd rather do this as a separate PR.

Reply:
> > Allowing replacing the open layer seems a bit sloppy to me. Previously, it was clear that each segment would have only one open layer at a time, now the rule is more complicated.
> 
> Do you see any issues with correctness here?

No. It's just a bit scary.

> We should definitely use better terminology here because now it is quite confusing.
> I think if we rename `open_layers` to `inmemory_layers`, things will be more clear - multiple in-memory layers can exist, but only one of them is open (aka writeable).
> But I'd rather do this as a separate PR.

Ok, works for me.